feat: default roles for user signups

zlwaterfield opened the pull request:
## Changes

Based on multiple user requests adding a default role for users when they join an organization. 
- new db fk to role on org model
- org serializer support to read/write
- UI to change the default role
- UI to show default role in roles table
- warming about clearing default role when deleting role
- invite.join logic to add user to role

Will we ever need support for multiple?

## How did you test this code?

- Need to review tests
- Need to look into auto provisioning joining org

zlwaterfield committed:
Add default role field to Organization model

Introduces a new 'default_role' field to the Organization model, allowing automatic assignment of a role to new members. Updates the serializer and adds the corresponding migration.

zlwaterfield committed:
Add is_default field to RoleSerializer

Introduces an is_default SerializerMethodField to RoleSerializer to indicate if a role is the organization's default role. The field is also set as read-only in the serializer's Meta configuration.

zlwaterfield committed:
Add default role selection for new organization members

Introduces UI and logic to set a default role for new members joining an organization. Updates organization type and logic to support the new 'default_role_id' property, and enhances role deletion flow to warn if the role is currently set as default.

zlwaterfield committed:
Auto-assign default role on organization membership

Adds logic to automatically assign a default role to a user when they join an organization, if a default role is configured. Exceptions during role assignment are captured for monitoring.

zlwaterfield committed:
Update activity log for default role

zlwaterfield committed:
Add user join tests for default role

zlwaterfield committed:
add org api test for default role

zlwaterfield committed:
fix new setting styling

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

yasen-posthog reviewed (approved):
Works like a charm!

Added a few things to double-check.

Generally, I find it very confusing for the Roles to be separated from Access control, and it took me a while to get my bearings and remember what the flow was. I bet it's 10x more confusing for non-techy people. Do you think it would be nice to consolidate everything under one page? Or at least copy the roles UI component on the access control page so they can manage the whole thing in one place.

zlwaterfield committed:
Add organization logic to role access controls

Integrates organizationLogic into roleAccessControlLogic and triggers organization loading when roles are updated. Also adds a disabledReason to the 'Add a role' button and sets a maxWidth for the delete role dialog for improved UI consistency.

zlwaterfield commented on posthog/models/user.py:
```diff
@@ -257,6 +258,26 @@ def join(
+            # Auto-assign default role if configured
+            if organization.default_role_id:
+                try:
+                    from ee.models import RoleMembership
```
Agree but we need to be careful about ee imports in posthog files.

zlwaterfield commented on posthog/models/user.py:
```diff
@@ -257,6 +258,26 @@ def join(
+                    RoleMembership.objects.create(
+                        role_id=organization.default_role_id, user=self, organization_member=membership
+                    )
+                except Exception as e:
```
I can check and move out of the transaction

zlwaterfield committed:
Improve default role assignment and activity log encoding

Refactored default role assignment in User model to occur outside of atomic transaction for better error handling. Enhanced RoleSerializer to handle missing or unauthenticated requests gracefully. Removed organization_id from activity log encoding for cleaner output.

zlwaterfield committed:
Merge remote-tracking branch 'origin/master' into zach/default-role

zlwaterfield committed:
Update 0829_organization_default_role.py

zlwaterfield committed:
Update 0829_organization_default_role.py

zlwaterfield committed:
Update 0829_organization_default_role.py

zlwaterfield committed:
Update test_organization.py

zlwaterfield committed:
Change how the guard works

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

zlwaterfield committed:
Merge branch 'master' into zach/default-role

zlwaterfield committed:
Merge branch 'master' into zach/default-role
